### DISCUSSION TITLE
tucker with sparse core tensor

### DISCUSSION THREAD

**Discussion by renjithravindran** [tensorly/tensorly#376] [Ideas]
Hi team Tensorly,

Is there any plan to add sparsity constraints to tucker decomposition?
I would like to have tucker with sparse core tensor.
If someone can show me how it can be done in the HOOI framework, i can give a try.

Thanks,
Renjith

**Comment by caglayantuna** [comment]
Hi Renjith, 

Tensorly already supports some sparse operations [Link](http://tensorly.org/stable/user_guide/sparse_backend.html). 

You can try Tucker decomposition with a sparse tensor as following example:

```python
import sparse
import numpy as np
import tensorly.contrib.sparse as stl
from tensorly.contrib.sparse.decomposition import tucker
from tensorly.contrib.sparse import tucker_to_tensor

coords = np.random.randint(0, 100, size=[3, 24])
data = list(range(1, 25))
X = sparse.COO(coords, data, shape=(100, 100, 100))
tensorly_tensor = stl.tensor(X, dtype='float')
core, factors = tucker(tensorly_tensor, rank=5, init='random')
res = tucker_to_tensor((core, factors))
```

**Reply by renjithravindran** [reply]
Hi caglayantuna,

What I am looking for is sparsity constraints on the factors not sparse input.
This is [available](http://tensorly.org/dev/auto_examples/decomposition/plot_guide_for_constrained_cp.html) for CP decomposition.
But not for tucker yet.

It would be great if someone can point me to a resource that shows how it could be done.

Thanks
Renjith

**Reply by cohenjer** [reply]
Hi @renjithravindran,
Great question ! Right now there is no support for constrained Tucker in Tensorly except nonnegative Tucker (which has some sparsity penalty on the core you can use, but I am unsure if it will perform correctly and we are working on it).
We may add constrained_tucker similar to how we have constrained CP, but that is not something that we are currently developing.
